File: my-first-project/gnews.js
#!/usr/bin/env node
import Parser from 'rss-parser';
import { parseArgs } from 'node:util';

const parser = new Parser();

// ANSI Colors
const RESET = "\x1b[0m";
const BOLD = "\x1b[1m";
const TITLE = "\x1b[1;37m";  // Bold White
const SOURCE = "\x1b[1;32m"; // Bold Green
const DATE = "\x1b[36m";     // Cyan
const LINK = "\x1b[4;34m";   // Underline Blue
const INDEX = "\x1b[1;33m";  // Bold Yellow
const HEADER = "\x1b[1;35m"; // Bold Magenta

const TOPICS = {
  world: 'WORLD',
  nation: 'NATION',
  business: 'BUSINESS',
  technology: 'TECHNOLOGY',
  entertainment: 'ENTERTAINMENT',
  sports: 'SPORTS',
  science: 'SCIENCE',
  health: 'HEALTH'
};

const options = {
  topic: { type: 'string', short: 't' },
  search: { type: 'string', short: 's' },
  limit: { type: 'string', short: 'l', default: '10' },
  help: { type: 'boolean', short: 'h' }
};

function printHelp() {
  console.log(`
Usage: node gnews.js [options]

Options:
  -t, --topic <name>    Filter news by topic.
                        Available topics: ${Object.keys(TOPICS).join(', ')}
  -s, --search <query>  Search articles by keyword
  -l, --limit <number>  Limit number of displayed news (default: 10)
  -h, --help            Show this help message
`);
}

async function main() {
  let values;
  try {
    const result = parseArgs({ options, allowPositionals: true });
    values = result.values;
  } catch (err) {
    console.error(`Error parsing arguments: ${err.message}`);
    printHelp();
    process.exit(1);
  }

  if (values.help) {
    printHelp();
    process.exit(0);
  }

  let url = 'https://news.google.com/rss?hl=en-US&gl=US&ceid=US:en';
  let titleContext = 'Top Stories';

  if (values.search) {
    const query = encodeURIComponent(values.search);
    url = `https://news.google.com/rss/search?q=${query}&hl=en-US&gl=US&ceid=US:en`;
    titleContext = `Search results for '${values.search}'`;
  } else if (values.topic) {
    const topicLower = values.topic.toLowerCase();
    if (TOPICS[topicLower]) {
      const topicId = TOPICS[topicLower];
      url = `https://news.google.com/rss/headlines/section/topic/${topicId}?hl=en-US&gl=US&ceid=US:en`;
      titleContext = `Topic: ${values.topic.toUpperCase()}`;
    } else {
      console.error(`Unknown topic '${values.topic}'. Available topics: ${Object.keys(TOPICS).join(', ')}`);
      process.exit(1);
    }
  }

  const limit = parseInt(values.limit, 10);
  if (isNaN(limit) || limit <= 0) {
    console.error(`Invalid limit value: ${values.limit}. Must be a positive integer.`);
    process.exit(1);
  }

  console.log(`Fetching news from Google (${titleContext})...`);

  try {
    const feed = await parser.parseURL(url);
    if (!feed.items || feed.items.length === 0) {
      console.log('No articles found.');
      return;
    }

    const count = Math.min(feed.items.length, limit);
    console.log(`\n${HEADER}=== Google News - Latest Articles (showing {count}) ===${RESET}`.replace('{count}', count));

    for (let i = 0; i < count; i++) {
      const item = feed.items[i];
      const title = item.title || 'No Title';
      const link = item.link || '';
      
      // Google News pubDate is usually a standard ISO or RFC822 string
      let dateStr = item.pubDate || '';
      if (dateStr) {
        try {
          const dt = new Date(dateStr);
          // Format as "Jun 15, 2026 07:57 PM"
          dateStr = dt.toLocaleString('en-US', {
            month: 'short',
            day: 'numeric',
            year: 'numeric',
            hour: '2-digit',
            minute: '2-digit',
            hour12: true
          });
        } catch (e) {
          // Keep raw string if parsing fails
        }
      }

      // Try to clean the title by removing the source suffix
      let cleanTitle = title;
      let sourceName = item.source?.title || '';
      
      if (!sourceName) {
        // Fallback: extract source from title (e.g. "Title - Source")
        const lastDash = title.lastIndexOf(' - ');
        if (lastDash !== -1) {
          cleanTitle = title.substring(0, lastDash);
          sourceName = title.substring(lastDash + 3);
        }
      } else {
        // If we have the source name explicitly, strip it from the title
        const suffix = ` - ${sourceName}`;
        if (cleanTitle.endsWith(suffix)) {
          cleanTitle = cleanTitle.slice(0, -suffix.length);
        }
      }

      console.log(`${INDEX}[${i + 1}]${RESET} ${TITLE}${cleanTitle}${RESET}`);
      console.log(`    Source: ${SOURCE}${sourceName || 'Unknown'}${RESET} | Date: ${DATE}${dateStr}${RESET}`);
      console.log(`    Link:   ${LINK}${link}${RESET}\n`);
    }
  } catch (err) {
    console.error(`Error fetching or parsing feed: ${err.message}`);
    process.exit(1);
  }
}

main();

File: my-first-project/gnews.py
#!/usr/bin/env python3
import urllib.request
import urllib.parse
import xml.etree.ElementTree as ET
import argparse
import sys
import os
import platform
import datetime

# Enable ANSI escape sequences on Windows
if platform.system() == 'Windows':
    os.system('')

# Force UTF-8 stdout if supported to prevent Unicode encoding errors on Windows
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')

# Colors
COLOR_RESET = "\033[0m"
COLOR_BOLD = "\033[1m"
COLOR_TITLE = "\033[1;37m"  # Bold White
COLOR_SOURCE = "\033[1;32m" # Bold Green
COLOR_DATE = "\033[36m"     # Cyan
COLOR_LINK = "\033[4;34m"   # Underline Blue
COLOR_INDEX = "\033[1;33m"  # Bold Yellow
COLOR_HEADER = "\033[1;35m" # Bold Magenta

TOPICS = {
    'world': 'WORLD',
    'nation': 'NATION',
    'business': 'BUSINESS',
    'technology': 'TECHNOLOGY',
    'entertainment': 'ENTERTAINMENT',
    'sports': 'SPORTS',
    'science': 'SCIENCE',
    'health': 'HEALTH'
}

def get_feed_url(topic=None, search=None):
    if search:
        query = urllib.parse.quote(search)
        return f"https://news.google.com/rss/search?q={query}&hl=en-US&gl=US&ceid=US:en"
    elif topic:
        topic_lower = topic.lower()
        if topic_lower in TOPICS:
            topic_id = TOPICS[topic_lower]
            return f"https://news.google.com/rss/headlines/section/topic/{topic_id}?hl=en-US&gl=US&ceid=US:en"
        else:
            print(f"Unknown topic '{topic}'. Available topics: {', '.join(TOPICS.keys())}")
            sys.exit(1)
    else:
        return "https://news.google.com/rss?hl=en-US&gl=US&ceid=US:en"

def fetch_rss(url):
    req = urllib.request.Request(
        url, 
        headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/115.0.0.0 Safari/537.36'}
    )
    try:
        with urllib.request.urlopen(req) as response:
            return response.read()
    except Exception as e:
        print(f"Error fetching news feed: {e}")
        sys.exit(1)

def parse_rss(xml_data):
    try:
        root = ET.fromstring(xml_data)
        items = []
        for item in root.findall('.//item'):
            title = item.find('title').text if item.find('title') is not None else 'No Title'
            link = item.find('link').text if item.find('link') is not None else ''
            pub_date = item.find('pubDate').text if item.find('pubDate') is not None else ''
            
            # Source is usually within <source> tag or in title (Google News appends " - Source Name" to title)
            source_tag = item.find('source')
            source = source_tag.text if source_tag is not None else ''
            
            # Clean title if it contains the source at the end
            clean_title = title
            if source and title.endswith(f" - {source}"):
                clean_title = title[:-len(f" - {source}")]
                
            items.append({
                'title': clean_title,
                'source': source,
                'link': link,
                'pub_date': pub_date
            })
        return items
    except Exception as e:
        print(f"Error parsing feed XML: {e}")
        sys.exit(1)

def display_news(items, limit):
    if not items:
        print("No articles found.")
        return
    
    count = min(len(items), limit)
    print(f"\n{COLOR_HEADER}=== Google News - Latest Articles (showing {count}) ==={COLOR_RESET}\n")
    
    for i in range(count):
        item = items[i]
        date_str = item['pub_date']
        try:
            # Parse RSS pubDate e.g., "Mon, 15 Jun 2026 21:00:00 GMT"
            dt = datetime.datetime.strptime(date_str, "%a, %d %b %Y %H:%M:%S %Z")
            date_str = dt.strftime("%b %d, %Y %I:%M %p")
        except Exception:
            pass
        
        print(f"{COLOR_INDEX}[{i+1}]{COLOR_RESET} {COLOR_TITLE}{item['title']}{COLOR_RESET}")
        print(f"    Source: {COLOR_SOURCE}{item['source']}{COLOR_RESET} | Date: {COLOR_DATE}{date_str}{COLOR_RESET}")
        print(f"    Link:   {COLOR_LINK}{item['link']}{COLOR_RESET}\n")

def main():
    parser = argparse.ArgumentParser(description="Fetch and display news from Google News RSS feeds.")
    group = parser.add_mutually_exclusive_group()
    group.add_argument("-t", "--topic", help=f"Filter news by topic. Available: {', '.join(TOPICS.keys())}")
    group.add_argument("-s", "--search", help="Search news articles by keyword")
    parser.add_argument("-l", "--limit", type=int, default=10, help="Limit number of displayed news (default: 10)")
    
    args = parser.parse_args()
    
    url = get_feed_url(topic=args.topic, search=args.search)
    
    title_context = "Top Stories"
    if args.search:
        title_context = f"Search results for '{args.search}'"
    elif args.topic:
        title_context = f"Topic: {args.topic.upper()}"
        
    print(f"Fetching news from Google ({title_context})...")
    xml_data = fetch_rss(url)
    items = parse_rss(xml_data)
    display_news(items, args.limit)

if __name__ == "__main__":
    main()
